src/routing/libosrm_wrapper.cpp

```diff
@@ -120,40 +122,85 @@ Matrices LibosrmWrapper::get_matrices(const std::vector<Location>& locs) const {
+  coords.reserve(locs.size());
 
-  // Ordering locations for the given steps, excluding
-  // breaks.
-  for (auto& step : route.steps) {
-    if (step.step_type != STEP_TYPE::BREAK) {
-      assert(step.location.has_value());
-      const auto& loc = step.location.value();
-      assert(loc.has_coordinates());
-      params.coordinates.emplace_back(osrm::util::FloatLongitude({loc.lon()}),
-                                      osrm::util::FloatLatitude({loc.lat()}));
-    }
+  for (const auto& loc : locs) {
+    coords.emplace_back(osrm::util::FloatLongitude({loc.lon()}),
+                        osrm::util::FloatLatitude({loc.lat()}));
   }
 
+  // Default options for routing.
+  osrm::RouteParameters
+    params(false, // steps
+           false, // alternatives
+           osrm::RouteParameters::GeometriesType::Polyline,
+           osrm::RouteParameters::OverviewType::Full,
+           false, // continue_straight,
+           std::move(coords),
+           std::vector<boost::optional<osrm::engine::Hint>>(),
+           std::vector<
+             boost::optional<double>>(coords.size(),
+                                      DEFAULT_LIBOSRM_SNAPPING_RADIUS));
+
   osrm::json::Object result;
   osrm::Status status = _osrm.Route(params, result);
 
   if (status == osrm::Status::Error) {
-    throw RoutingException(
-      result.values["code"].get<osrm::json::String>().value + ": " +
-      result.values["message"].get<osrm::json::String>().value);
+    throw_error(result, locs);
   }
 
   auto& result_routes = result.values["routes"].get<osrm::json::Array>();
-  auto& json_route = result_routes.values.at(0).get<osrm::json::Object>();
+  return std::move(result_routes.values.at(0).get<osrm::json::Object>());
+}
+
+void LibosrmWrapper::update_sparse_matrix(
+  const Id v_id,
+  const std::vector<Location>& route_locs,
+  Matrices& m,
+  std::mutex& matrix_m,
+  std::unordered_map<Id, std::string>& v_id_to_geom,
+  std::mutex& id_to_geom_m) const {
+  auto json_route = get_route_with_coordinates(route_locs);
+
+  auto& legs = json_route.values["legs"].get<osrm::json::Array>();
+  assert(legs.values.size() == route_locs.size() - 1);
+
+  for (std::size_t i = 0; i < legs.values.size(); ++i) {
+    auto& leg = legs.values.at(i).get<osrm::json::Object>();
+
+    std::scoped_lock<std::mutex> lock(matrix_m);
+    m.durations[route_locs[i].index()][route_locs[i + 1].index()] =
+      utils::round<UserDuration>(
+        leg.values["duration"].get<osrm::json::Number>().value);
+    m.distances[route_locs[i].index()][route_locs[i + 1].index()] =
+      utils::round<UserDistance>(
+        leg.values["distance"].get<osrm::json::Number>().value);
+  }
+
+  std::scoped_lock<std::mutex> lock(id_to_geom_m);
+  v_id_to_geom.try_emplace(v_id,
+                           std::move(json_route.values["geometry"]
+                                       .get<osrm::json::String>()
+                                       .value));
```

Comment:
Storing geometries from first route requests in a `std::unordered_map` based on vehicle id is not a good idea since technically having vehicles with identical ids is allowed.